### PR TITLE
Nucal Helper Functions

### PR DESCRIPTION
**Pull request by tyler-a-cox**
This PR implements two new functions for `nucal`, `build_nucal_wgts` and `compute_spatial_filters`. `build_nucal_wgts` is a simple wrapper on top of `build_data_wgts` and implements additional functionality such as flagging certain u-modes and spectral window ranges. `compute_spatial_filters` uses `hera_filters` to compute sets of PSWF basis functions for filtering the spatial axis

### CODE REVIEW

File: hera_cal/nucal.py [hunk 1]
```diff
@@ -464,3 +495,147 @@ def __iter__(self):
...
+def compute_spatial_filters(radial_reds, freqs, ell_half_width=1, eigenval_cutoff=1e-12, cache={}):
+    """
+    Compute prolate spheroidal wave function (PSWF) eigenvectors filters for each radially redundant group in radial_reds. 
+    Note that if you are using a large array with a large range short and long baselines in an individual radially
+    redundant group, it is advised to filter radial_reds using radial_reds.filter_reds before running this function 
+    to reduce the size of filters generated
+
+    Parameters:
+    ----------
+    radial_reds : FrequencyRedundant object
```
[the remark this answers is not in the document]
Comment (jsdillon) [reply hera_cal/nucal.py:508]: Also, I feel like we're mixing the names of radial reds and FrequencyRedundancy objects, no?

File: hera_cal/nucal.py [hunk 2]
```diff
@@ -464,3 +495,147 @@ def __iter__(self):
...
+                wgts[:, umag < min_u_cut] = True
+            if max_u_cut is not None:
+                wgts[:, umag > max_u_cut] = True
+            if min_freq_cut is not None:
+                wgts[:, freqs < min_freq_cut] = True
+            if max_freq_cut is not None:
+                wgts[:, freqs > max_freq_cut] = True
+            if spw_range_flags is not None:
+                for spw in spw_range_flags:
+                    wgts[:, (freqs > spw[0]) & (freqs < spw[1])] = True
```
Comment (jsdillon) [inline hera_cal/nucal.py:626]: Shouldn't these `True`s be 0s?

**Review by jsdillon — CHANGES_REQUESTED** [review]
A handful of comments and possible issues, but mostly it's about getting the names of things and documentation as clear as possible and as self-consistent as possible.

File: hera_cal/nucal.py [hunk 3]
```diff
@@ -1,11 +1,40 @@
...
+import warnings
 import numpy as np
 from copy import deepcopy
+from hera_filters import dspec
 import astropy.constants as const
 from scipy.cluster.hierarchy import fclusterdata
 
+
+# Optional import of Optax and Jax libraries
+try:
```
[the remark this answers is not in the document]
Comment (tyler-a-cox) [reply hera_cal/nucal.py:16]: Just added it

File: hera_cal/nucal.py [hunk 4]
```diff
@@ -1,11 +1,40 @@
...
+    import optax
+
+    # Approved Optax Optimizers
+    OPTIMIZERS = {
+        'adabelief': optax.adabelief, 'adafactor': optax.adafactor, 'adagrad': optax.adagrad, 'adam': optax.adam,
+        'adamw': optax.adamw, 'fromage': optax.fromage, 'lamb': optax.lamb, 'lars': optax.lars,
+        'noisy_sgd': optax.noisy_sgd, 'dpsgd': optax.dpsgd, 'radam': optax.radam, 'rmsprop': optax.rmsprop,
+        'sgd': optax.sgd, 'sm3': optax.sm3, 'yogi': optax.yogi
```
[the remark this answers is not in the document]
Comment (tyler-a-cox) [reply hera_cal/nucal.py:24]: This is just left over from some of the functions I originally had in this PR, but removed. I'll take it out for now and add it back in another PR.

File: hera_cal/nucal.py [hunk 5]
```diff
@@ -219,10 +248,12 @@ def __init__(
         self.blvec_error_tol = blvec_error_tol
 
         if reds is None:
-            reds = redcal.get_reds(antpos, pols=pols, bl_error_tol=bl_error_tol)
+            self.reds = redcal.get_reds(antpos, pols=pols, bl_error_tol=bl_error_tol)
+        else:
+            self.reds = reds
```
[the remark this answers is not in the document]
Comment (tyler-a-cox) [reply hera_cal/nucal.py:253]: Added a short test for this

File: setup.py [hunk 6]
```diff
@@ -61,11 +61,13 @@ def package_files(package_dir, subdirectory):
...
         'aipy',
     ],
-    # 'extras_require': {
-    #     "all": [
-    #         'aipy @ git+https://github.com/hera-team/aipy'
-    #     ]
-    # },
+    'extras_require': {
+         "all": [
+             'aipy @ git+https://github.com/hera-team/aipy',
```
[the remark this answers is not in the document]
Comment (tyler-a-cox) [reply setup.py:66]: Removed this

File: hera_cal/nucal.py [hunk 7]
```diff
@@ -464,3 +495,147 @@ def __iter__(self):
     def sort(self, key=None, reverse=True):
         """Sorts list by length of the radial groups"""
         self._radial_groups.sort(key=(len if key is None else key), reverse=reverse)
+
+def compute_spatial_filters(radial_reds, freqs, ell_half_width=1, eigenval_cutoff=1e-12, cache={}):
+    """
+    Compute prolate spheroidal wave function (PSWF) eigenvectors filters for each radially redundant group in radial_reds. 
+    Note that if you are using a large array with a large range short and long baselines in an individual radially
```
[the remark this answers is not in the document]
Comment (tyler-a-cox) [reply hera_cal/nucal.py:502]: Fixed this

File: hera_cal/nucal.py [hunk 8]
```diff
@@ -1,11 +1,40 @@
...
+
+try:
+    import jax
+    from jax import numpy as jnp
+    jax.config.update("jax_enable_x64", True)
+
+except:
+    warnings.warn('Jax is not installed. Some functionality may not be available')
+    
+# Constants
```
[the remark this answers is not in the document]
Comment (tyler-a-cox) [reply hera_cal/nucal.py:37]: We did. That must have been left over from the previous PR. I changed the comment

File: hera_cal/nucal.py [hunk 9]
```diff
@@ -464,3 +495,147 @@ def __iter__(self):
...
+    Compute prolate spheroidal wave function (PSWF) eigenvectors filters for each radially redundant group in radial_reds. 
+    Note that if you are using a large array with a large range short and long baselines in an individual radially
+    redundant group, it is advised to filter radial_reds using radial_reds.filter_reds before running this function 
+    to reduce the size of filters generated
+
+    Parameters:
+    ----------
+    radial_reds : FrequencyRedundant object
+        FrequencyRedundant object containing lists of radially redundant baselines. 
+    freqs : np.ndarray
```
[the remark this answers is not in the document]
Comment (tyler-a-cox) [reply hera_cal/nucal.py:510]: Changed the name of the object, so I don't think that this is necessary now.

File: hera_cal/nucal.py [hunk 10]
```diff
@@ -464,3 +495,147 @@ def __iter__(self):
...
+        Frequencies found in the data in units of Hz
+    ell_half_width : float, default=1
```
[the remark this answers is not in the document]
Comment (tyler-a-cox) [reply hera_cal/nucal.py:512]: Changed it to `spatial_half_width`

File: hera_cal/nucal.py (hunk 2 above)
Comment (tyler-a-cox) [reply hera_cal/nucal.py:626]: Changed the name of the variable to `flags` so this shouldn't be a problem now

File: hera_cal/__init__.py [hunk 11]
```diff
@@ -34,3 +34,4 @@
 from . import autos
 from . import noise
 from . import tempcal
+from . import nucal
```
Comment (jsdillon) [inline hera_cal/__init__.py:37]: It's good practice to put a newline at the end of a file

**Review by jsdillon — APPROVED** [review]
One tiny change, otherwise good to go.

File: hera_cal/__init__.py (hunk 11 above)
Comment (tyler-a-cox) [reply hera_cal/__init__.py]: Oh, didn't know that, thanks for letting me know. Just added that change.
